Change: Company manager face definitions and dynamic UI.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Company manager faces are brittle and have no room for extension. The language used can also be problematic, e.g. "European" and "African".

Bitstuffed values are used to control sprite selections, with lots of special casing to handle differences between male and female faces. Validation of company face bits does not always take account of all cases, which can allow invalid faces to be used in some cases.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Adds a face definition system. This definition is similar to the old CompanyManagerFaceBitsInfo data, but instead of trying to describe all 4 faces in one definition, splits the default faces into 4 separate styles

Each face style can have a variable number of layers, and this is even used by the default faces, removing the special cases used before. Variables can also include toggles and palette selections which affect other variables, used for things like glasses, earrings, moustache, and eye colour.

The user interface is now dynamic and only shows the layers relevant to the current style and toggleable options. This is drawn in a matrix widget, in a style similar to other settings windows.

For future-proofing, each face style is given a string label, so that the numerical index of the style is not important. There is no NewGRF support in this PR, but that could be developed later.

Bumps the savegame version as face style is stored separately.

Simple face selection window now has only a "New Face" button to change the face. This randomises all variables AND the style:
![image](https://github.com/user-attachments/assets/6acdbaea-5bfe-413e-b1fb-2e2ddf5d4b1d)

Advanced face selection, with dynamic rendering of face options.
![image](https://github.com/user-attachments/assets/38ebc574-ac93-4c3a-8316-6f42b96ee248)

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

The AI/GS script API includes functions to get and set the company president gender. The switch from 2 gender and 2 ethnicity options to a 4 separate face styles consciously removes the distinction of gender and ethnicity, so the API functions fake this based on the index.

Savegame changes might want to store face style by label instead of index, if NewGRF faces are introduced.

Some code parts are a bit TOO far along the line of adding NewGRF support, and can be toned back a bit.

Lacking doxygen.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

2TallTyler commented on src/table/settings/misc_settings.ini:
```diff
@@ -268,13 +268,11 @@ min      = 1
-def      = 0
-min      = 0
-max      = 0xFFFFFFFF
+def      = """"
```
Extra quotes?

```suggestion
def      = ""
```

PeterN replied:
Yes, this is correct for settings.

2TallTyler commented on src/saveload/saveload.h:
```diff
@@ -403,6 +403,8 @@ enum SaveLoadVersion : uint16_t {
 	SLV_FIX_SCC_ENCODED_NEGATIVE,           ///< 353  PR#14049 Fix encoding of negative parameters.
 	SLV_ORDERS_OWNED_BY_ORDERLIST,          ///< 354  PR#13948 Orders stored in OrderList, pool removed.
 
+	SLV_FACE_STYLES,                        ///< 355  PR#..... Addition of face styles, replacing gender and ethnicity.
```
```suggestion
	SLV_FACE_STYLES,                        ///< 355  PR#14319 Addition of face styles, replacing gender and ethnicity.
```

2TallTyler commented on src/company_type.h:
```diff
@@ -49,7 +49,11 @@ class CompanyMask : public BaseBitSet<CompanyMask, CompanyID, uint16_t> {
+
+struct CompanyManagerFace {
+	uint style = 0; ///< Company manager face style.
+	uint32_t bits = 0; ///< Company manager face bits, info see in company_manager_face.h
```
Typo or maybe I misunderstood. 🙂 

```suggestion
	uint32_t bits = 0; ///< Company manager face bits, see company_manager_face.h
```

PeterN replied:
Just an old comment written by a non-native speaker.

2TallTyler reviewed (commented):
In general, I like this approach. Haven't dug too deeply into the code, but a few things I noticed.

* The advanced face selection window is missing a resize handle.
* The `Player face no.` input contains text by default but the input field only lets me type numbers. The error string for an invalid input is also incorrect.
* It might be good to change the documentation for `ScriptCompany::GetPresidentGender` to say that the function works but is depreciated.
* You mention maybe wanting to store faces by label instead of index, for future NewGRF support. I have no technical comments, but I do think future NewGRF support is a good idea.

2TallTyler commented on src/company_cmd.cpp:
```diff
@@ -1375,3 +1384,147 @@ CompanyID GetFirstPlayableCompanyID()
+/**
+ * Mask company manager face bits to ensure they are all within range.
+ * @note Does not update the CompanyManagerFace itself. Unused bits are cleared.
+ * @param cmf The CompanyMnaagerFace.
```
```suggestion
 * @param cmf The CompanyManagerFace.
```

2TallTyler commented on src/company_cmd.cpp:
```diff
@@ -1375,3 +1384,147 @@ CompanyID GetFirstPlayableCompanyID()
+}
+
+/**
+ * Set a companny face style.
```
```suggestion
 * Set a company face style.
```

2TallTyler reviewed (commented):
Feedback based on latest push:

* The new resize handle can't drag a window taller (if the scrollbar shows extra room).
* `Player face no.` input works fine, error string is a bit vague but I don't know what a better one would say.
* When I mentioned depreciating gender for scripts, I meant [here](https://github.com/OpenTTD/OpenTTD/blob/master/src/script/api/script_company.hpp#L185-L200) as well as the changelog. Not sure if your comments in `script_company.cpp` were intended to address this comment, if not disregard. 🙂 
* Some functions still missing Doxygen, let me know if you want me to flag them.

2TallTyler commented on src/company_cmd.cpp:
```diff
@@ -1052,11 +1059,13 @@ CommandCost CmdCompanyAllowListCtrl(DoCommandFlags flags, CompanyAllowListCtrlAc
  * @return the cost of this operation or an error
  */
-CommandCost CmdSetCompanyManagerFace(DoCommandFlags flags, CompanyManagerFace cmf)
+CommandCost CmdSetCompanyManagerFace(DoCommandFlags flags, uint32_t bits, uint style)
```
Doxygen needs updating to new params.

2TallTyler reviewed (commented):
This looks good to me, I'm willing to YOLO approve once Doxygen is added.

I have no technical opinion on savegame storage (label vs index) for future GRF spec extension, but am not overly concerned about making it absolutely bulletproof if players do things like upgrade GRF versions, change GRFs in a running game, etc. It's just a face. Do with that what you will. 🙂 

Edit: cmcaine and I posted at the same time, I agree with that. It would be nice for the glasses selection to sit next to the glasses toggle. 😄

PeterN committed:
Change: New company face definition system.

Bits used by company faces are now defined by a variable system instead of being hardcoded, allowing future expansion.

The four face types covering gender and skin colour are now separate face styles with their own definitions.

2TallTyler commented on src/table/company_face.h:
```diff
@@ -0,0 +1,98 @@
+	{FaceVarType::Sprite, 12, 31, 1, 2, SpriteID{0x347}, STR_FACE_GLASSES},
+};
+
+/** Variables of first feminine face. */
```
```suggestion
/** Variables of first female face. */
```

2TallTyler reviewed (commented):
New ordering of options is really nice. Sorry for the one nitpick for consistency, otherwise looks good to me. 🙂
